Fix illegal modification of artifact queue

lukpueh opened the pull request:
**Fixes issue #**:
None

**Description of the changes being introduced by the pull request**:
During artifact rule verification `verify_item_rules` may pass the
`source_artifacts_queue` to `verify_match_rule`, which returns an
updated version of the queue (minus matched artifacts) and
overrides the original `source_artifacts_queue`.

However, the queue is passed to `verify_match_rule` by reference,
and modified directly. While this wouldn't be a problem for
`source_artifacts_queue`, as it is modified and then returned to
override itself, there is another reference pointing to that list,
i.e. (depending on whether `verify_item_rules` operates on material
or product rules) `source_products_queue` or `source_artifacts_queue`,
neither of which should not be modified by `verify_match_rule`.

This commit changes `verify_match_rule` to not modify the passed
reference and also makes an actual copy of `source_materials_queue`
and `source_products_queue` in `verify_item_rules` just to be safe.

Please fill in the fields below to submit a pull request.  The more information
that is provided, the better.


**Please verify and check that the pull request fulfills the following
requirements**:

- [ ] The code follows the [Code Style Guidelines](https://github.com/secure-systems-lab/code-style-guidelines#code-style-guidelines)
- [ ] Tests have been added for the bug fix or new feature
- [ ] Docs have been added for the bug fix or new feature

Commit:
Fix illegal modification of artifact queue

During artifact rule verification `verify_item_rules` may pass the
`source_artifacts_queue` to `verify_match_rule`, which returns an
updated version of the queue (minus matched artifacts) and
overrides the original `source_artifacts_queue`.

However, the queue is passed to `verify_match_rule` by reference,
and modified directly. While this wouldn't be a problem for
`source_artifacts_queue`, as it is modified and then returned to
override itself, there is another reference pointing to that list,
i.e. (depending on whether `verify_item_rules` operates on material
or product rules) `source_products_queue` or `source_artifacts_queue`,
neither of which should not be modified by `verify_match_rule`.

This commit changes `verify_match_rule` to not modify the passed
reference and also makes an actual copy of `source_materials_queue`
and `source_products_queue` in `verify_item_rules` just to be safe.

Commit:
Refactor artifact rule verification (WIP) [ci skip]

It turns out that there are more problems in our
artifact rule verification than addressed in
4b3028aab6b83823b1c8c78b7a8e4d9244b5e554

The biggest problem is probably that it is very convoluted, i.e.
there are different "queues" that share the same lists in memory
and get modified under different conditions, which inevitably
leads (lead) to bugs.

This commit trys to align more closely with the pseudo-code
in the spec (which OTOH is a little bit too oversimplified, e.g.
there is only one queue, which IMHO is not enough).

TODO:
 - Try to simplify even more
 - Update docstrings
 - Fix tests / add better tests

SantiagoTorres commented on in_toto/verifylib.py:
```diff
@@ -1137,64 +1105,47 @@ def verify_item_rules(source_name, source_type, rules, links):
+
+    # Unreachable
+    else: # pragma: no cover
+      pass
```
Would it be useful to log this in case it's really not unreachable? (I know, I know)

lukpueh replied:
I still don't know you to universally best handle these. I once triggered are related discussion but we never agreed on a best practice (see https://github.com/secure-systems-lab/code-style-guidelines/issues/18).

lukpueh replied:
But yes, pass is a bad idea. Maybe raise a custom `UnreachableError`? :)

SantiagoTorres commented on in_toto/verifylib.py:
```diff
@@ -889,16 +879,13 @@ def verify_delete_rule(rule, source_materials_queue, source_products_queue):
+      source_products_queue, rule_data["pattern"])
 
-  return list(set(source_materials_queue) - set(matched_materials))
+  return set(filtered_materials) - set(filtered_products)
```
I'm still surprised everywhere else this is treated like a set and nothing happened...

lukpueh replied:
Not sure I understand. Before, all `verify_*_rule` functions returned the updated queue(s) as `list` (except for `disallow`). Then I discovered that `match` also modified the queue at the passed reference, which is a bug that I fixed in 4b3028aab6b83823b1c8c78b7a8e4d9244b5e554.

And then I realized that there are other things in the way the queues are updated that need fixing, which I've started in bf0709bb96f3f5f2eeac8723c1146b627bdc84c5, in the course of which I'm trying to simplify the code and align it more with the pseudo code of the spec, e.g. having `verify_*_rule` return the consumed artifacts (as `set`) and update the queue(s) on the caller side.

The reason why we haven't noticed is probably that it requires very specific combinations and sequences of rules and artifact queues to become an issue (and probably because the code is confusing).

I'll follow up with a more precise description of the issue and thorough tests.

SantiagoTorres reviewed (approved):
This mostly LGTM. I know it's a draft but the changeset looks reasonable enough.

Commit:
More refactor artifact rule verification (WIP) [ci skip]

- Remove `source_` prefix from variable names for less clutter
 (is this necessary, or does this just clutter the diff?)
- Start working comments
- Make use of reference thing

lukpueh commented:
See #261 and #262 for more details and a full fix of the here drafted issue.
